File: Algorithms/Sorting_Algorithms/Merge_Sort/C/Merge_Sort.c
//Program in C to impement merge sort algorithm
#include<stdio.h>

//Iterative function to merge individual list items into a list
//It returns the list sorted in ascending order
int merge(int array[],int lb,int mid,int ub)
{
    int i=lb,k=lb;
    int newarr[20];
    int j=mid+1;
    //for comparison of array[i] with array[j]
    while(i<=mid&&j<=ub)
    {
        if(array[i]<array[j])
        {
            newarr[k]=array[i];
            i++;
        }
        else
        {
            newarr[k]=array[j];
            j++;
        }
        k++;
    }
    //if array[i] gets exhausted,add array[j] list items
    if(i>mid)
    {
        newarr[k]=array[j];
        j++;
        k++;
    }
    else
    {
        //if array[j] gets exhausted,add array[i] list items
        while(i<=mid)
        {
            newarr[k]=array[i];
            i++;
            k++;
        }
    }
    //adding elements to the sorted list
    for(k=lb; k<=ub; k++)
    {
        array[k]=newarr[k];
    }
}

//divide list into halves recursively
int mergeSort(int array[],int lb,int ub)
{
    if(lb<ub)
    {
        int mid=(lb+ub)/2;
        //dividing lists until a single list item remains
        mergeSort(array,lb,mid);
        mergeSort(array,mid+1,ub);
        //merging into a complete sorted list
        merge(array,lb,mid,ub);
    }
}

//Utility function to print the sorted array
int printArray(int array[],int lb,int ub)
{
    printf(" The required array after Sorting :\n");
    for(int i=lb; i<=ub; i++)
    {
        printf(" %d ",array[i]);
    }
}

//Driver program to check the above functions with sample input
int main()
{
    int array[]= {33,5,997,-4,0,18,7,-58};
    int lb=0;
    int ub=7;
    int mid=(lb+ub)/2;
    mergeSort(array,lb,ub);
    printArray(array,lb,ub);
    return 0;
}

File: Algorithms/Sorting_Algorithms/Merge_Sort/Java/Merge_Sort.java
package sorting;

public class MergeSortMain {

  public static void main(String[] args) {
    MergeSort obj = new MergeSort();
    int a[] = {7, -9, 0, 4, 8, 9, 1, 15, 44, 2, -4, 6};

    System.out.println("");
    obj.mergesort(a, 0, a.length - 1);
    obj.printArray(a);
  }
}
class MergeSort {
  public void mergesort(int arr[], int l, int r) {
    if (r > l) {
      int m = (l + r) / 2;
      mergesort(arr, l, m); // sort the two array
      mergesort(arr, m + 1, r);
      merge(arr, l, m, r); // merge the array
    }
  }
  // function to merge the array
  public void merge(int arr[], int l, int m, int r) {
    int leftArray[] = new int[m - l + 1];
    int rightArray[] = new int[r - m];

    // storing the array in left and right
    for (int i = 0; i < m - l + 1; i++) {
      leftArray[i] = arr[l + i];
    }
    for (int j = 0; j < r - m; j++) {
      rightArray[j] = arr[m + 1 + j];
    }
    int i = 0, j = 0;
    int k = l;
    while (i < m - l + 1 && j < r - m) {
      if (leftArray[i] <= rightArray[j]) {
        arr[k] = leftArray[i];
        i++;
      } else {
        arr[k] = rightArray[j];
        j++;
      }
      k++;
    }
    while (i < m - l + 1) { // if any element remaining in left array
      arr[k] = leftArray[i];
      i++;
      k++;
    }
    while (j < r - m) {
      arr[k] = rightArray[j]; // if any element remaining in right array
      k++;
      j++;
    }
  }

  // printing the sorted array
  public void printArray(int[] array) {
    for (int i = 0; i < array.length; i++) {
      System.out.print(array[i] + " ");
    }
  }
}
